Added a state property for the Request class

FirstKlaas opened the pull request:
Added a `state` property for the `responder.models.Request` class. The property just exposes the starlette state property. No extra logic added. No testcase written. Added a short docstring to the property.

taoufik07 commented:
Thanks, can you add some simple tests ?

FirstKlaas commented:
Ok. Had to read a quick pytest tutorial first. Never used it before. Seems pretty straight forward. Why do you have no makefile and requirements file (for development), no linting... Did I miss something?

iancleary commented:
This project uses `black` for linting: https://github.com/taoufik07/responder/blob/master/.travis.yml#L17 and pipenv for virtual environments and requirements.
  
The other thing I have found is app.py and app2.py are in the .gitignore file as well https://github.com/taoufik07/responder/blob/master/.gitignore#L16-L17 in case you need

https://python-responder.org/en/latest/testing.html (possibly could use some updates per this process)

Not sure if I missed anything, @taoufik07 ?

taoufik07 commented:
in addition to what @iancleary said, we're using `pipenv` so our requirements file is the Pipfile and Pipfile.lock.

For dev purposes you need to use `pipenv install --dev` in order to install dev dependencies or if you don't use `pipenv` go to the Pipfile grab the packages names and install them.

@FirstKlaas if you have some hard time writing the tests I can assist you

taoufik07 commented:
The test can be something like 

```python
def test_request_state(api): 
     @api.route("/")      
     async def home(req, resp): 
         req.state.username = "Klaas" 
         resp.text = "First {}".format(req.state.username) 

     r = api.requests.get("/")
     assert r.text == "First Klaas"
```

FirstKlaas commented:
Hi, the "pipenv" info was the point I was missing. Interesting. I added a test case. Thanks for the answers to my question. So I think

FirstKlaas commented:
Great @taoufik07. My test is slightly different, as I added a minimal middelware. I think this is more realistic.

taoufik07 commented on tests/test_request_state.py:
```diff
@@ -0,0 +1,30 @@
+class StateMiddleware(BaseHTTPMiddleware):
+
+    async def dispatch(self, request, call_next):
+        request.state.test1 = int(12)
```
no need to call the int

taoufik07 commented on tests/test_request_state.py:
```diff
@@ -0,0 +1,30 @@
+
+
+def test_api_request_state(api, url):
+    
```
can you move the middleware inside the test function

taoufik07 commented on tests/test_request_state.py:
```diff
@@ -0,0 +1,30 @@
+        assert hasattr(req.state, "test1")
+        assert hasattr(req.state, "test2")
+        assert isinstance(req.state.test1, int)
+        assert isinstance(req.state.test2, str)
```
`assert request.state.test1 == 12` and `assert request.state.test2 == "foo"` are enough no need for the `hasattr` or `isinstance`

taoufik07 replied:
I don't think having the tests inside the home view is a good idea, you can simply return the state values in `resp.text` and do the test

taoufik07 commented on Pipfile:
```diff
@@ -15,6 +15,7 @@ flask = "*"
 sphinx = "*"
 marshmallow = "*"
 pytest-cov = "*"
+pylint = "*"
```
can you please remove `pylint` ?

taoufik07 commented on Pipfile:
```diff
@@ -4,7 +4,7 @@ verify_ssl = true
 
 [packages]
-responder = {editable = true, path = "."}
+responder = {editable = true,path = "."}
```
and add the space after the comma

FirstKlaas commented:
Ok. I removed the asserts from the home view. Although I think it was a good idea. Just to make shure, that no type conversion happen during chaining. But I removed them anyway. I ran black. Now everything should be fine. I'm happy when this will be released, as I'm working on a larger plattform and decided to go with responder. Great library so far. Keep up the good work.

FirstKlaas commented:
Again sorry. Overlooked your request to move the test. Wait a second.

taoufik07 commented:
@FirstKlaas thanks for contribution !

taoufik07 commented:
Can you squash the commits to keep the log tree clean before I merge ?

taoufik07 commented:
You can use something like `git reset --soft HEAD~6 && git commit -m "Add state for requests"`

taoufik07 commented:
Just try with `git reset --soft HEAD~8 && git commit -m "Add state for requests"` and then force push `git push -f origin iss382-state-for-request` so we end up with a clean reflog

FirstKlaas commented:
Now i ruined everything. :( Now it says Pipfile.lock is conflicted. But locally I have no conflicts. I just copied the two git commands you wrote, as I don't fully understand, what they are doing. I would be easier to create a new branch, do the clean changes and then create a new pull request. Otherwise I have no clue what to do.

taoufik07 commented:
@FirstKlaas ok xD

FirstKlaas commented:
Hi @taoufik07, created a new branch with a clean implementation, testcase and documentation. Created a new pull request based on the new branch. I will delete this branch and you can fire and forget this pull request. Thanks for your patience and your support. I learned two new things, which is great. pipenv and pytest. The latter I knew, but never used. The first was totally new to my. Next contribution will be smoother. ;)

taoufik07 commented:
closed via #385
